src/charm.py

```diff
@@ -1,303 +1,312 @@
+        certs = gen_certs(model=self._namespace, service_name=f"{self._service_name}")
+        for k, v in certs.items():
+            setattr(self._stored, k, v)
+
+    def _upload_certs_to_container(self):
+        """Upload generated certs to container."""
         try:
-            image_details = self.image.fetch()
-        except OCIImageResourceError as e:
-            raise CheckFailed(f"{e.status.message}", e.status_type)
-        return image_details
+            self._check_container_connection(self.container)
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
+
+        self.container.push(CONTAINER_CERTS_DEST + "/key.pem", self._cert_key, make_dirs=True)
+        self.container.push(CONTAINER_CERTS_DEST + "/cert.pem", self._cert, make_dirs=True)
+
+    def _check_container_connection(self, container: Container) -> None:
+        """Check if connection can be made with container.
+        Args:
+            container: the named container in a unit to check.
+        Raises:
+            ErrorWithStatus if the connection cannot be made.
+        """
+        if not container.can_connect():
+            raise ErrorWithStatus("Pod startup is not complete", MaintenanceStatus)
+
+    def _on_install(self, _):
+        """Installation only tasks."""
+        # deploy K8S resources to speed up deployment
+        self._apply_k8s_resources()
+
+    def _on_remove(self, _):
+        """Remove all resources."""
+        delete_error = None
+        self.unit.status = MaintenanceStatus("Removing K8S resources")
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        crd_resources_manifests = self.crd_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, k8s_resources_manifests)
+        except ApiError as error:
+            # do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Failed to delete CRD resources, with error: {error}")
+                delete_error = error
+        try:
+            delete_many(self.crd_resource_handler.lightkube_client, crd_resources_manifests)
+        except ApiError as error:
+            # do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Failed to delete K8S resources, with error: {error}")
+                delete_error = error
+
+        if delete_error is not None:
+            raise delete_error
+
+        self.unit.status = MaintenanceStatus("K8S resources removed")
+
+    def _on_pebble_ready(self, event):
+        """Configure started container."""
+        try:
+            self._check_container_connection(self.container)
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
```

Comment:
you already do check inside upload certs
```suggestion
```

Reply:
Even though comment is correct, in context of Pebble function we don't know that `_upload_certs_to_container` checks for connection. The connection check here is more to protect eventual `_on_event` call at the end of `_on_pebble_ready` function.
Upload certs function can be re-used in other places, like upgrade.
This is how it was done in Seldon around this https://github.com/canonical/seldon-core-operator/blob/main/src/charm.py#L352
If there is a better solution let's discuss it.
Either way, LGTM.

Reply:
on event does not require container

and if it does, then we need to put it into the function that needs it

Reply:
Making it a point to discuss when we need to check container connection.